src/main/kotlin/org/jitsi/nlj/rtp/bandwidthestimation/BandwidthEstimator.kt

```diff
@@ -0,0 +1,241 @@
+     * @param[seq] A 16-bit sequence number of packets processed by this
+     *  [BandwidthEstimator].
+     */
+    fun processPacketLoss(now: Instant, sendTime: Instant?, seq: Int) {
+        if (timeSeriesLogger.isTraceEnabled) {
+            val point = diagnosticContext.makeTimeSeriesPoint("bwe_packet_loss", now)
+            if (sendTime != null) {
+                point.addField("sendTime", sendTime.formatMilli())
+            }
+            point.addField("seq", seq)
+            timeSeriesLogger.trace(point)
+        }
+
+        doProcessPacketLoss(now, sendTime, seq)
+    }
+
+    /**
+     * A subclass's implementation of [processPacketLoss].
+     *
+     * See that function for parameter details.
+     */
+    protected abstract fun doProcessPacketLoss(now: Instant, sendTime: Instant?, seq: Int)
+
+    /**
+     * Inform the bandwidth estimator about a new round-trip time value
+     */
+    fun onRttUpdate(now: Instant, newRtt: Duration) {
+        if (timeSeriesLogger.isTraceEnabled) {
+            val point = diagnosticContext.makeTimeSeriesPoint("bwe_rtt", now)
+            point.addField("rtt", newRtt.formatMilli())
+            timeSeriesLogger.trace(point)
+        }
+
+        doRttUpdate(now, newRtt)
+    }
+
+    /**
+     * A subclass's implementation of [onRttUpdate].
+     *
+     * See that function for parameter details.
+     */
+    protected abstract fun doRttUpdate(now: Instant, newRtt: Duration)
+
+    /** Get the estimator's current estimate of the available bandwidth.
+     *
+     * @param[now] The current time, when this function is called.
+     */
+    abstract fun getCurrentBw(now: Instant): Bandwidth
+
+    /** Get the current statistics related to this estimator.
+     *
+     * @param[now] The current time, when this function is called.
+     */
+    abstract fun getStats(now: Instant): NodeStatsBlock
+
+    /** Reset the estimator to its initial state. */
+    abstract fun reset(): Unit
+
+    interface Listener {
+        fun bandwidthEstimationChanged(newValue: Bandwidth)
+    }
+
+    private val listeners = LinkedList<Listener>()
+    private var curBandwidth = (-1).bps
+
+    private var lastBweLogTime = NEVER
+    private val minBweLogInterval = Duration.ofMillis(500)
+
+    /**
+     * Notifies registered listeners that the estimate of the available
+     * bandwidth has changed.
+     */
+    @Synchronized
+    protected fun reportBandwidthEstimate(now: Instant, newValue: Bandwidth) {
+        if (timeSeriesLogger.isTraceEnabled) {
+            if (newValue != curBandwidth ||
+                lastBweLogTime == NEVER ||
+                Duration.between(lastBweLogTime, now) >= minBweLogInterval) {
+                val point = diagnosticContext.makeTimeSeriesPoint("bwe_estimate", now)
+                point.addField("bw", newValue.bps)
+                timeSeriesLogger.trace(point)
+                lastBweLogTime = now
+            }
+        }
+
+        if (newValue == curBandwidth) {
```

Comment:
should this allow some wiggle room?  or is that handled elsewhere?  i guess it makes sense to let handlers decide what threshold makes sense for them.

Reply:
Yeah, I figured we'd let handlers decide how much change is interesting.